civiform/civiform
# Remove applicant id from URLs for index action

### Description

Currently, routes for applicant actions include the applicant id in the URL. We want to remove these for applicant actions to provide cleaner URLs for sharing and bookmarking. To accomplish this, we will store the applicant id in the user profile. A profile attribute is set in `ProfileFactory.createNewApplicant()` which will achieve the goal for every newly-created profile. If a user with an existing (legacy) profile authenticates successfully, then we also set the applicant id in the `CiviFormProfileMerger`.

This PR only affects the `ApplicantProgramsController.index()` action; subsequent PRs will address the other applicant actions.

This PR renames the existing `index()` method to `indexWithApplicantId()`. A new `index()` method extracts the applicant ID from the profile and then calls `indexWithApplicantId()`. The corresponding `routes` are configured appropriately.

Note that we want to retain the existing routes when _TIs_ are taking actions of an applicant's behalf.  Also, some applicants will have existing profiles that do not yet contain the applicant id, so we need to use the existing routes for them as well. Since routes for applicant actions are computed in a number of views as well as controllers (for redirects), the new `ApplicantRoutes` class will centralize the conditional logic for computing the appropriate route for a given situation. There are also some changes in plumbing to pass the user profile to views when this computation is required. Note that another reason to retain the existing routes, particularly for the home page, is that existing users may have the current URLs bookmarked.

In addition to the unit tests, this change was tested on a dev instance with an existing profile in the browser cookie. See instructions below.

### Release notes

Remove applicant ID from home page URL

### Checklist

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)


### Instructions for manual testing

- In a dev instance, check out `main`
- Navigate to `http://localhost:9000`
- This will redirect to the home page with the applicant ID in the URL. It will also create a user profile that does not include the applicant id
- Check out this branch
- Again, navigate to `http://localhost:9000`
- Again, this should redirect to the home page with the applicant ID in the URL
- Click "End session" in upper right. This will create a new user profile that _does_ include the applicant id
- Verify that the browser is now on `http://localhost:9000/programs` (no applicant id)

### Issue(s) this completes

Relates to #5905
